### PR TITLE
[128] - TTS functionality for AAQ

### PR DESCRIPTION
**Pull request by MustafaAkolawala**
Reviewer: @lickem22 
Estimate: 20 mins

---

## Ticket

Fixes:  [Ticket link](https://idinsight.atlassian.net/browse/AAQ-588?atlOrigin=eyJpIjoiNTlmY2M3OThjMTViNGIzNDg1ZmU3NzViMTE0NDI1NDciLCJwIjoiaiJ9)

## Description

This PR introduces text-to-speech (TTS) functionality to AAQ's ` llm-response` endpoint, enhancing user accessibility by allowing responses to be delivered in audio format as well as text. A reusable TTS component, `generate_speech` has been created to facilitate this functionality across different parts of the application.

### Goal

The primary goal of this PR is to increase the accessibility of the AAQ platform. By integrating TTS capabilities, users who prefer auditory information can now listen to responses generated by the language model. This functionality ensures that AAQ is more inclusive and caters to a broader audience.

### Changes

- **New Directory Creation:** A new directory named voice_api has been created to house components related to STT and TTS functionality, ensuring better organization and modularity.

- **Endpoint Modification:** The llm-response endpoint has been modified to include TTS functionality. This involves converting text responses to speech and providing a downloadable MP3 file as part of the response, also the `QueryResponse` pydantic model got modified by adding a `tts_file`  as optional to it.

- **Reusable Component:** A reusable TTS component called generate_speech has been implemented. This component takes text input and generates an MP3 file using the Google Text-to-Speech (gTTS) API.

- **Testing:** Unit tests and integration tests have been created and updated to ensure the new TTS functionality works as intended and does not disrupt existing features.

## How has this been tested?

dev-environment
docker-compose
Pytests

Steps to Test the TTS functionality

1. **Enable TTS Functionality:** Provide a flag in `query_metadata` with `"use_tts": true` to enable TTS functionality. 
2. **Generate TTS Response:** Make a request to the llm-response endpoint with the TTS flag enabled. The TTS file will be generated and stored in the ` core_backend` Docker container.
3. **Retrieve the TTS File:** Use the following command in your terminal to copy the file from the Docker container to your local repository root:
`docker cp <container_name/id>:/usr/src/aaq_backend/<name.mp3> ./<name.mp3>`
 Replace <container_name/id> with the appropriate container name or ID, and <name.mp3> with the name of the generated MP3 file.
4. **Verify the TTS File:** Once the file is copied to your local repository, you can play the MP3 file to verify the TTS functionality.

## Checklist

Fill with `x` for completed. 

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [x] I have updated the automated tests

### CODE REVIEW

File: core_backend/app/question_answer/routers.py
Comment (lickem22) [inline core_backend/app/question_answer/routers.py:147]: We should pass `use_tts` in the schema and not inside the metadata

Comment (MustafaAkolawala) [reply core_backend/app/question_answer/routers.py:147]: so in the `QueryBase` schema just add another optional flag right for `use_tts`?

File: core_backend/tests/api/test_question_answer.py
Comment (lickem22) [inline core_backend/tests/api/test_question_answer.py:253]: Since you are using it twice, put it in `conftest.py`

File: core_backend/tests/api/test_question_answer.py
Comment (lickem22) [inline core_backend/tests/api/test_question_answer.py:359]: You shouldn't print the JSON response

**Review by lickem22 — CHANGES_REQUESTED** [review]
Thanks for this Mustadfa. I left a few comments but main things to change are:
- Put STT logic within `question_answer`
- Refactor a few tests
- I try testing STT locally but I am getting this:
```
"state": "error",
  "tts_file": "response_1.mp3"
```
Is there something I am missing? Like an api key to use gTTS ? 
I am happy to jump on a call to discuss that.

File: core_backend/tests/api/test_question_answer.py
Comment (MustafaAkolawala) [reply core_backend/tests/api/test_question_answer.py:371]:
```
if expect_tts_file:
    assert "tts_file" in json_response
    assert json_response["tts_file"] is not None
 else:
     assert "tts_file" not in json_response
```

does something like this makes sense?

File: core_backend/app/question_answer/routers.py
Comment (lickem22) [reply core_backend/app/question_answer/routers.py:147]: Yes

File: core_backend/tests/api/test_question_answer.py
Comment (lickem22) [reply core_backend/tests/api/test_question_answer.py:371]: Yes that works!

File: core_backend/app/question_answer/models.py
Comment (lickem22) [inline core_backend/app/question_answer/models.py:137]: You don't need to do that since you gave a default value to `tts_file` in the pydantic model.

File: core_backend/tests/api/test_TTS.py
Comment (lickem22) [inline core_backend/tests/api/test_TTS.py:37]: You should add another use case when you type gibberish and the language is supported. Also I think it's worth checking when the text is in a language (like hindi) and the specified language is different (for ex Swahili) to see the behaviour.

File: core_backend/app/question_answer/routers.py
Comment (lickem22) [inline core_backend/app/question_answer/routers.py:145]: I just thought about something, in the audio support flow, the audio we are supposed to return is not the question asked by the user but the answer by the LLM, so I suggest we don't modify the question answer module for now but later after we have the full audio support logic. Sorry for not flagging this earlier. There are two ways you can go about that since you are not merging to main:
- Deleting every code you updated in `question_answer` to update it later when needed
- only delete this section for now and return None for the `tts_file`until you implement that logic. 
I leave it up to you how you want to do it. But in both cases, you should remove the changes you made in `test_question_answer`.

**Review by lickem22 — APPROVED** [review]
Thanks Mustafa, 
I only have one comment about returning the question as an audio in the response. I will leave it up to you how to update it. 
Also, it is probably not. something to worry about now, but how do we handle errors when the TTS function returns an audio but the only thing said in the audio is `error`. I thought in case of an error, an exception will be raised. I will approve this PR but please implement the changes I made before merging.

File: core_backend/tests/api/test_TTS.py
Comment (MustafaAkolawala) [reply core_backend/tests/api/test_TTS.py:37]: it would not raise any error right now for these specific test cases cause: 
1) as the TTS is being used fairly late in the workflow (after we get a response from the LLM) the guardrails would already catch errors and not let the TTS file get generated at all.
2) as the external TTS we using right now is GTTS, it is limited in it capabilities and cannot identify if the langauge would be gibberish or there is a language mismatch.

but i think that should not be an issue right as when used with the rest of the workflow and with the necessary guardrails implemented (as we would mostly get the correct language identified and it wouldnt give an answer to gibberish) these wouldnt pose a risk, but we can still try implementing extra error handling in the inhouse TTS

Comment (lickem22) [reply core_backend/tests/api/test_TTS.py:37]: It makes sense not to test TTS doing that on question_answer, but here we are testing the TTS function alone, so We should make sure the function is robust enough to catch these issues.

Comment (MustafaAkolawala) [reply core_backend/tests/api/test_TTS.py:37]: so what do u suggest, right now the external API is not robust enough to catch these errors but when we replace it with an actual neural model we could catch these errors at that time so for now should i merge it?

Comment (lickem22) [reply core_backend/tests/api/test_TTS.py:37]: If the api don't catch it then the test wont pass. You can merge for now

**Pull request merged**
